### ISSUE TITLE
Error: In computing Tumour Subclusters

### ISSUE THREAD

**Issue by reshu23** [broadinstitute/infercnv#174]
Hi,

I am getting following error, when I am running inferCNV::run, 

analysis mode:"subclusters",
cutoff=0.1,
denoise=TRUE,
HMM=TRUE,
num_threads=10,
tumor_subcluster_pval=0.05
                            
STEP 05: computing tumor subclusters via random_trees
INFO [2019-09-18 07:58:11] define_signif_tumor_subclusters(p_val=0.05
INFO [2019-09-18 07:58:11] ::subtract_ref_expr_from_obs:Start inv_log=TRUE, use_bounds=TRUE
INFO [2019-09-18 07:58:11] subtracting mean(normal) per gene per cell across all data
INFO [2019-09-18 07:58:16] -subtracting expr per gene, use_bounds=TRUE
INFO [2019-09-18 07:58:21] -mirroring for hspike
INFO [2019-09-18 07:58:21] ::subtract_ref_expr_from_obs:Start inv_log=TRUE, use_bounds=TRUE
INFO [2019-09-18 07:58:21] subtracting mean(normal) per gene per cell across all data
INFO [2019-09-18 07:58:22] -subtracting expr per gene, use_bounds=TRUE
INFO [2019-09-18 07:58:22] define_signif_tumor_subclusters(), tumor: all_observations
unique tumor clade name: all_observations.1
INFO [2019-09-18 08:39:12] random trees, using 10 parallel threads
scheduled cores 2, 4 did not deliver results, all values of the jobs will be affectedError in .parameterize_random_cluster_heights_smoothed_trees(tumor_expr_data,  : 
  (list) object cannot be coerced to type 'double'

**Comment by GeorgescuC** [comment]
Hi @reshu23 ,

How many cells does your dataset have? And how much memory did your machine/server have available when you ran inferCNV?
So far, when people had errors with "scheduled cores ___ did not deliver results", the issues have been due to too many threads compared to available resources. I would try to reduce the num_threads value and see if the error still occurs.

Regards,
Christophe.

**Comment by reshu23** [comment]
Hi,
Thank you.  I have 6343 cells in observation. I have just checked, my machine has 8 processor. And memory is:
MemTotal:       31961360 kB
MemFree:         7697400 kB
MemAvailable:   28114292 kB

I have also tried num_threads=8. I have got same error. Now, I am going to use default value, which is 4 thread. I will let you know, If I get same error again.

**Comment by reshu23** [comment]
Sorry, I have 11113 cells in observation and 6343 cells in Reference

**Comment by reshu23** [comment]
Hi,

Finally, I have got success(subcluster computation). In the cell grouping file: I have information of 8 sub clusters:

all_observations.all_observations.1.1.1.1
all_observations.all_observations.1.1.1.2
all_observations.all_observations.1.1.2.1
all_observations.all_observations.1.1.2.2
all_observations.all_observations.1.2.1.1
all_observations.all_observations.1.2.1.2
all_observations.all_observations.1.2.2.1
all_observations.all_observations.1.2.2.2

How can I correlate this cluster number with inferCNV heatmap observation partition? Should I count in heatmap, from top to bottom as all_observations.all_observations.1.1.1.1, all_observations.all_observations.1.1.1.2 and so on.
Thank you.

**Comment by GeorgescuC** [comment]
Hi @reshu23 ,

On the heatmap, the clusters are ordered from bottom to top compared to the order written in the legend for the observations. You can also get the mapping from subcluster to cells by reading the infercnv_obj@tumor_subclusters slot that contains the list of cell indices for each subcluster (indices refer to the cell order in the infercnv_obj@expr.data matrix).

Regards,
Christophe.

**Comment by reshu23** [comment]
Thank you. I think, I am little bit confused. My observation cells has only one group. I found subcluster "all_observations.all_observations.1.2.2.1" at the bottom. I recognized it by cnv region info(by looking at heatmap) and small group of cells. 

We can extract cell indices for each subcluster from infercnv_obj as well as from text file "HMM_CNV_predictions.HMMi6.rand_trees.hmm_mode-subclusters.Pnorm_0.5.cell_groupings.txt". Right?. They must be same. I will check it with infercnv_obj@tumor_subcluster slot.

Actually, I am trying to understand subclusters in heatmap. So, I can say, this region in the heatmap is subcluster so and so. File: "infercnv.references.txt", this text file contains clustered reference cells matrix(clustered cells) as it has been shown in the heatmap (references). Am I right?
